hack/build.sh

```diff
@@ -44,7 +89,12 @@ MODE="${MODE:-release}"
 # Build terraform binaries before setting environment variables since it messes up make
 if test "${SKIP_TERRAFORM}" != y && ! (echo "${TAGS}" | grep -q -e 'aro' -e 'altinfra')
 then
-  make -j8 -C terraform all
+  if check_terraform_providers; then
```

Comment:
Q: now that we have a working triggering condition in the [terraform variant job](https://github.com/openshift/release/blob/f4f1bf0fe4f6af1eb6525431de841db8ba9a22b8/ci-operator/config/openshift/installer/openshift-installer-master__terraform.yaml#L51), and thus the `installer-terraform-providers` image is rebuilt and promoted when a change is detected in the `terraform/` path, is there any additional benefit in keeping this check here? Ie, in case the `SKIP_TERRAFORM` was set, the current job will fetch the latest `installer-terraform-providers` image available from the configured is, for example `ocp:4.15`

Reply:
The benefit is two-fold:
1) importing the binaries from the image results in their modification time being older than the go.{mod,sum}, causing `make` to think they need to be rebuilt. `check_terraform_providers` uses the `go version` module information to determine if the binaries need rebuilding;
2) for the cases where we *are* updating the providers in a PR, this check will detect that and force their rebuild. That is needed because we cannot build and promote the `installer-terraform-providers` in the same job (promotion is a post-submit step).

Reply:
Thanks @r4f4 for clarifying! After a second though, and based on the points above, what about moving the `make -j8 -C terraform all` as additional test directly in the [terraform variant](https://github.com/openshift/release/blob/ac7eae7ff72ec05171dc8804ef57236c5c7a74a6/ci-operator/config/openshift/installer/openshift-installer-master__terraform.yaml#L43) (and only if the change is in the terraform path or go.mod - and still reusing `check_module_changes()`)?
In this way the terraform variant job will always act as a _producer_ for `installer-terraform-providers` in case of change, while the other jobs will always act as consumers, fetching and using the latest `installer-terraform-providers` available.
The second point could be addressed by splitting the PR in two, ie by first updating/bumping the providers (via the variant job) and then using them (via the normal job), as it happens usually when the dependency is placed into another different repo. IIUC it's not a very common operation, so in this manner all the PRs that are not interested in the terraform artifacts will not perform any check and take just the latest image.

Reply:
@andfasano how's that different from the current behavior? Isn't the image already produced and promoted when there is a change to the providers, since the generation of the image is a first step in the terraform variant job? I think I don't really understand the benefit of that approach.

Reply:
I think @zaneb was against a 2-step bump process, that's why we went with this build-when-needed approach. If we're going to do a 2-step process, I'd rather use a separate repo (which was the idea in https://github.com/openshift/installer/pull/7621).

Reply:
After a clarification with @r4f4 it looks like that the current implementation is good enough to cover all the required use cases ( change not in tf path / change in tf path /  change in and otuside the tf path), so I'm fine to proceed with it